src/lib/statistics/statistics_objects/abstract_histogram.cpp

```diff
@@ -0,0 +1,1035 @@
+        if (search_prefix_bin != INVALID_BIN_ID && search_prefix_next_value_bin != INVALID_BIN_ID &&
+            bin_height(search_prefix_bin) == 0u &&
+            (bin_height(search_prefix_next_value_bin) == 0u ||
+             bin_minimum(search_prefix_next_value_bin) == search_prefix_next_value)) {
+          for (auto current_bin = search_prefix_bin + 1; current_bin < search_prefix_next_value_bin; current_bin++) {
+            if (bin_height(current_bin) > 0u) {
+              return false;
+            }
+          }
+          return true;
+        }
+
+        return false;
+      }
+
+      return false;
+    }
+    case PredicateCondition::NotLike: {
+      if (!LikeMatcher::contains_wildcard(*value)) {
+        return _does_not_contain(PredicateCondition::NotEquals, variant_value);
+      }
+
+      // If the pattern starts with a MatchAll, we can only prune it if it matches all values.
+      if (value->front() == '%') {
+        return *value == "%";
+      }
+
+      /**
+       * We can also prune prefix searches iff the domain of values captured by the histogram is less than or equal to
+       * the domain of strings captured by a prefix pattern.
+       *
+       * Example:
+       * min: car
+       * max: crime
+       * predicate: col NOT LIKE 'c%'
+       *
+       * With the same argument we can also prune predicates in the form of 'c%foo',
+       * where foo can be any pattern itself.
+       * We only have to consider the pattern up to the first MatchAll character.
+       */
+      const auto match_all_index = value->find('%');
+      if (match_all_index != pmr_string::npos) {
+        const auto search_prefix = value->substr(0, match_all_index);
+        if (search_prefix == bin_minimum(BinID{0}).substr(0, search_prefix.length()) &&
+            search_prefix == bin_maximum(bin_count() - 1).substr(0, search_prefix.length())) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+    default: {
+      auto value_in_domain = _domain.string_to_domain(*value);
+
+      std::optional<AllTypeVariant> value2_in_domain;
+      if (variant_value2) {
+        value2_in_domain = boost::get<pmr_string>(*variant_value2);
+      }
+
+      return _general_does_not_contain(predicate_condition, value_in_domain, value2_in_domain);
+    }
+  }
+}
+
+template <typename T>
+CardinalityAndDistinctCountEstimate AbstractHistogram<T>::estimate_cardinality_and_distinct_count(
+    const PredicateCondition predicate_condition, const AllTypeVariant& variant_value,
+    const std::optional<AllTypeVariant>& variant_value2) const {
+  auto value = static_variant_cast<T>(variant_value);
+  if (!value) {
+    return {static_cast<Cardinality>(total_count()), static_cast<float>(total_distinct_count())};
+  }
+
+  if constexpr (std::is_same_v<
+                    T,
+                    pmr_string>) {  // NOLINT clang-tidy is crazy and sees a "potentially unintended semicolon" here...
```

Comment:
format

Reply:
Is this even still the case with the new clang version?

Reply:
thankfully apparently not